Parachute plugin for simulating flight termination in gazebo

Jaeyoung-Lim opened the pull request:
This PR adds a parachute plugin which is used to simulate the flight termination in gazebo sitl.

The parachute model the plugin comes from https://github.com/AurelienRoy/ardupilot_sitl_gazebo_plugin but integrated into the px4 sitl_gazebo

Since the original code is distributed under Apache, I am not sure if there needs to be a more correct license header in the parachute plugin. The code itself needed to be modified quite a lot to be able to work nicely with the px4 HIL interface.

Below is the demo of the parachute plugin working with the roll angle exceeding the limit.

Video can be seen in the link: [Video](https://youtu.be/17x3Y7K8-j0)

TSC21 commented on models/parachute_small/model.config:
```diff
@@ -0,0 +1,18 @@
+  <name>Parachute small</name>
+  <version>1.0</version>
+  <sdf version="1.5">model.sdf</sdf>
+
```
You can add the license as:
```suggestion
  <license>Apache License, Version 2.0</license>
```

TSC21 commented on include/gazebo_parachute_plugin.h:
```diff
@@ -0,0 +1,88 @@
+  double terminate_rot_vel_ = -1700;
+  int motor_number_;
+
+  std::string trigger_sub_topic_ = "~/video_stream";
```
Why is the trigger the `video_stream` topic?

TSC21 commented on include/gazebo_parachute_plugin.h:
```diff
@@ -0,0 +1,88 @@
+
+  bool attached_parachute_ = false;
+  double max_rot_velocity_ = 3500;
+  double ref_motor_rot_vel_;
```
```suggestion
  double ref_motor_rot_vel_ = 0;
```

TSC21 commented on include/gazebo_parachute_plugin.h:
```diff
@@ -0,0 +1,88 @@
+  physics::WorldPtr world_;
+  event::ConnectionPtr _updateConnection;
+
+  bool attached_parachute_ = false;
```
Can you add some comments on what each one of these variables is doing? Please use `///<`.

TSC21 commented on include/gazebo_parachute_plugin.h:
```diff
@@ -0,0 +1,88 @@
+  virtual void OnUpdate(const common::UpdateInfo&);
+
+private:
+  void LoadParachute();
```
Some comments here can be used as well:
Ex:
```c++
  /// \brief Trigger the parachute
  /// \param _msg msg received to trigger the parachute
  void TriggerCallback(const boost::shared_ptr<const msgs::Int> &_msg);
```

TSC21 commented on models/if750a/if750a.sdf:
```diff
@@ -494,67 +499,13 @@
-          <input_index>7</input_index>
-          <input_offset>0</input_offset>
-          <input_scaling>0.524</input_scaling>
+          <input_scaling>1000</input_scaling>
```
Can you explain the rationale behind this?

Jaeyoung-Lim replied:
@TSC21 I just wanted to use the same scale as other channels. Was this the question?

TSC21 commented on src/gazebo_parachute_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+void ParachutePlugin::AttachParachute(physics::ModelPtr &parachute_model){
+
+#if GAZEBO_MAJOR_VERSION >= 9
+  const ignition::math::Pose3d uavPose = model_->WorldPose();
```
Can we change the naming from `uavPose` to `vehicle_pose`? Also, the naming format doesn't seem to be uniform (some camel case, then some snake case). Let's keep the variables snake case, unless you have another suggestion.

Jaeyoung-Lim replied:
Done!

TSC21 reviewed (changes requested):
@Jaeyoung-Lim just some minor stuff. Otherwise good to go.

Jaeyoung-Lim committed:
Add parachute plugin


Load parachute plugin


Add parachute model


Parachute fixes


Fix joint


wip parachute trigger


Add trigger topic


Subscribe to motor command speeds


Fix


Subscribe to motor channel topic


Trigger parachute in flight termination state


Cleanup 1


Attach parachute


Fix


Add license statement


Address comments


Handle case for gazebo7


Fix


More


Update include/gazebo_parachute_plugin.h

Co-Authored-By: Nuno Marques <n.marques21@hotmail.com>
Rename uavPose


Address comments


f

Jaeyoung-Lim committed:
Clear formatting and cleanups

Includes clearing up the code by clearing up misused subscripts and making repeated patterns into a function
